docs/proposals/20210524-etcdadm-bootstrap-provider.md

```diff
@@ -0,0 +1,582 @@
+
+  - Cluster API supports the use of an external etcd cluster, by allowing users to provide their external etcd cluster's endpoints.
+  So it would be good to add support for provisioning the etcd cluster too, so the user can have a single workflow for full management of a cluster using the external etcd topology. The user can offload the responsibility of managing an external etcd cluster to CAPI.
+    
+  - External etcd topology decouples the control plane and etcd member. So if a control plane-only node fails, or if there's a memory leak in a component like kube-apiserver, it won't directly impact an etcd member.
+    
+  - Etcd is resource intensive, so it is safer to have dedicated nodes for etcd, it could use more disk space, higher bandwidth. 
+    Having a separate etcd cluster for these reasons could ensure a more resilient HA setup.
+    
+- Motivation behind using etcdadm as a pluggable etcd provider
+  
+  - Leveraging existing projects such as etcdadm is one way of bringing up an etcd cluster. One of the goals mentioned in the [etcdadm project roadmap](https://github.com/kubernetes-sigs/etcdadm/blob/master/ROADMAP.md) is to integrate it with CAPI to support the "external etcd" use case.
+  
+  - Etcd providers can be made pluggable and we can start by adding one that uses etcdadm.
+
+### Goals
+
+- Introduce pluggable etcd providers in Cluster API, starting with an etcdadm based provider. This etcdadm provider will create and manage an etcd cluster.
+- User should be able to create a Kubernetes cluster that uses external etcd topology in a single step.
+- User should only have to create the required CAPI resources, along with the required etcd provider specific resources. That should trigger creation of an etcd cluster, followed by creation of the target workload cluster which uses this etcd cluster. Meaning, the CAPI controllers will deploy the etcd cluster first and then use that as input to the workload cluster, without requiring any user intervention.
+- There will be a 1:1 mapping between the external etcd cluster and the workload cluster.
+- Define a contract for pluggable etcd providers and define steps that control plane providers should take to use a managed external etcd cluster.
+- Support the following etcd cluster management actions: scale up and scale down, etcd member replacement and etcd version upgrades.
+- The etcd providers will utilize the existing Machine objects to represent etcd members for convenience instead of adding a new machine type for etcd.
+
+### Non-Goals
+- The first iteration will use IP addresses/hostnames as etcd cluster endpoints. It can not configure static endpoint(s) till the Load Balancer provider is available.
+- API changes such as adding new fields within existing control plane providers, like the KubeadmControlPlane provider. We will utilize the existing external etcd endpoints field from the KubeadmConfigSpec for KCP.
```

Comment:
This is tricky.
KCP is a pluggable component in CAPI, thus we should not make assumptions it exist in a Cluster, and technically the only assumption we can make on a control plane provider are documented in https://cluster-api.sigs.k8s.io/developer/architecture/controllers/control-plane.html#contracts.

That's means "We will utilize the existing external etcd endpoints field from the KubeadmConfigSpec for KCP." can't be used or the scope of this proposal must be changed so we are targeting only cluster with KCP (but this is really limiting)

Reply:
the intention was to say that we won't add any new fields to the existing control plane provider API types specifically for external etcd clusters. And I wanted to include KCP only as an example. I have rephrased this in an attempt to make that clear. Does that look better now or should I remove this, because I agree that changing the scope of this proposal to target KCP only is limiting.

Reply:
Modified the proposal goals to make this external etcd provider work only with KCP for now. Added the proposed change in control plane contract in the non-goals/future work section so we can treat it as a separate proposal